HPXMLtoOpenStudio/measure.rb

```diff
@@ -2433,109 +2395,51 @@ def self.add_hot_water_and_appliances(runner, model, weather, spaces)
         end
         solar_fraction = 0.0 if solar_fraction.nil?
 
-        ec_adj = HotWaterAndAppliances.get_dist_energy_consumption_adjustment(@has_uncond_bsmnt, @cfa, @ncfl,
-                                                                              dist_type, recirc_control_type,
-                                                                              pipe_r, std_pipe_length, recirc_loop_length)
-
-        runner.registerInfo("EC_adj=#{ec_adj}") # Pass value to tests
+        ec_adj = HotWaterAndAppliances.get_dist_energy_consumption_adjustment(@has_uncond_bsmnt, @cfa, @ncfl, hot_water_distribution)
 
         dhw_load_frac = water_heating_system.fraction_dhw_load_served * (1.0 - solar_fraction)
 
-        @dhw_map[sys_id] = []
+        @dhw_map[water_heating_system.id] = []
 
-        if wh_type == HPXML::WaterHeaterTypeStorage
+        if water_heating_system.water_heater_type == HPXML::WaterHeaterTypeStorage
 
-          tank_vol = water_heating_system.tank_volume
-          re = water_heating_system.recovery_efficiency
-          capacity_kbtuh = water_heating_system.heating_capacity / 1000.0
+          Waterheater.apply_tank(model, loc_space, loc_schedule, water_heating_system, ec_adj, @dhw_map, desuperheater_clg_coil, solar_fraction)
 
-          Waterheater.apply_tank(model, loc_space, loc_schedule, fuel, capacity_kbtuh, tank_vol,
-                                 ef, re, setpoint_temp, ec_adj, @dhw_map,
-                                 sys_id, desuperheater_clg_coil, jacket_r, solar_fraction)
+        elsif water_heating_system.water_heater_type == HPXML::WaterHeaterTypeTankless
 
-        elsif wh_type == HPXML::WaterHeaterTypeTankless
+          Waterheater.apply_tankless(model, loc_space, loc_schedule, water_heating_system, ec_adj, @nbeds, @dhw_map, desuperheater_clg_coil, solar_fraction)
 
-          performance_adjustment = water_heating_system.performance_adjustment
+        elsif water_heating_system.water_heater_type == HPXML::WaterHeaterTypeHeatPump
 
-          Waterheater.apply_tankless(model, loc_space, loc_schedule, fuel, ef, performance_adjustment,
-                                     setpoint_temp, ec_adj, @nbeds, @dhw_map,
-                                     sys_id, desuperheater_clg_coil, solar_fraction)
+          Waterheater.apply_heatpump(model, runner, loc_space, loc_schedule, weather, water_heating_system, ec_adj, @dhw_map, desuperheater_clg_coil, solar_fraction, @living_zone)
 
-        elsif wh_type == HPXML::WaterHeaterTypeHeatPump
+        elsif (water_heating_system.water_heater_type == HPXML::WaterHeaterTypeCombiStorage) || (water_heating_system.water_heater_type == HPXML::WaterHeaterTypeCombiTankless)
 
-          tank_vol = water_heating_system.tank_volume
+          combi_sys_id_list << water_heating_system.id
```

Comment:
The sequence of calling apply_combi, then HotWaterAndAppliances.apply(), finally Waterheater.apply_combi_system_EMS is the reason, because the EMS program will want water use equipment information. Though not sure if the list is the best way to resolve it.

Reply:
I see. Well how about getting rid of this list and changing the apply_combi_system_EMS method to operate on all systems? I.e., pass in @hpxml.water_heating_systems and, inside the method, loop through all the systems to get the combi system ids.

Then we just have a single line of code in measure.rb -- the call to Waterheater.apply_combi_system_EMS.